### PR TITLE
[EG-440] Add some error codes and the error resource generation tool

### PR DESCRIPTION
**Pull request by JamesHR3**
## Overview
This PR adds a few more error codes and also introduces some utilities for working with error codes.

The main addition is a resource generation tool that creates the required resource bundle files given a set of locales and a JAR file containing the built error code definitions. This helps to avoid issues where required resources and properties are not generated correctly.

Due to an issue discovered late on, the UI has changed slightly to this tool, and now the fully qualified name of the error code classes is also required:
```
java -jar tools/error-tool/build/libs/corda-tools-error-utils-4.5-SNAPSHOT.jar generate-resources node/build/libs/corda-node-4.5-SNAPSHOT.jar common/logging/src/main/resources/error-codes/ "net.corda.node.utilities.logging.TestErrors"

```

## Testing
 - Verified that the `generate-resources` command correctly generates resources for missing codes, and does nothing for present codes
- Ensured a sensible error is produced if the specified class doesn't exist in the provided JAR
- Manual testing of the new error codes

### CODE REVIEW

File: common/logging/src/test/kotlin/net/corda/commmon/logging/errorReporting/ErrorCodeTest.kt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+ */
+abstract class ErrorCodeTest<T>(private val clazz: Class<T>,
+                                private val printProperties: Boolean = false) where T: Enum<T>, T: ErrorCodes {
+
+    abstract val dataForCodes: Map<T, List<Any>>
+
+    private class TestError<T>(override val code: T,
+                               override val parameters: List<Any>) : ErrorCode<T> where T: Enum<T>, T: ErrorCodes
+
+    @Test(timeout = 300_000)
```
Comment (fenryka) [inline common/logging/src/test/kotlin/net/corda/commmon/logging/errorReporting/ErrorCodeTest.kt:32]: these seem very long for a simple test (assuming this is 30 minutes?

File: tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+    companion object {
+        private const val ERROR_INFO_RESOURCE = ".*ErrorInfo.*"
+
+        /**
+         * List all resource bundle names in a given directory
+         */
+        fun listResourceNames(location: Path) : List<String> {
+            return location.toFile().walkTopDown().filter {
+                it.name.matches("[^_]*\\.properties".toRegex()) && !it.name.matches(ERROR_INFO_RESOURCE.toRegex())
```
Comment (fenryka) [inline tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt]: probably best to only compile the regex once then just use it, but I assume this tool isn't run more than once so it probably doesn't matter,

File: tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+            return location.toFile().walkTopDown().filter {
+                it.name.matches(".*\\.properties".toRegex())
+            }.map { it.name }.toList()
+        }
+
+        /**
+         * Create a classloader with all URLs in a given directory
+         */
+        fun loaderFromDirectory(location: Path) : URLClassLoader {
+//            val urls = location.toFile().walkTopDown().map { it.toURI().toURL() }.asIterable().toList().toTypedArray()
```
Comment (fenryka) [inline tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt]: debug?

File: tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorTool.kt [hunk 4]
```diff
@@ -0,0 +1,31 @@
+package net.corda.errorUtilities
+
+import net.corda.cliutils.CordaCliWrapper
+import net.corda.cliutils.ExitCodes
+import net.corda.cliutils.start
+import net.corda.errorUtilities.docsTable.DocsTableCLI
+import net.corda.errorUtilities.resourceGenerator.ResourceGeneratorCLI
+
+fun main(args: Array<String>) {
+    val tool = ErrorTool()
```
Comment (fenryka) [inline tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorTool.kt]: not super idiomatic kotlin

**Review by fenryka — CHANGES_REQUESTED** [review]

File: tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt (hunk 3 above)
Comment (JamesHR3) [reply tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt]: Ah missed that - thanks!

File: tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt (hunk 2 above)
Comment (JamesHR3) [reply tools/error-tool/src/main/kotlin/net/corda/errorUtilities/ErrorResourceUtilities.kt]: I've made all these `private val` - makes it slightly clearer in the code what we're matching against.

File: common/logging/src/test/kotlin/net/corda/commmon/logging/errorReporting/ErrorCodeTest.kt (hunk 1 above)
Comment (JamesHR3) [reply common/logging/src/test/kotlin/net/corda/commmon/logging/errorReporting/ErrorCodeTest.kt:32]: Seems to be our default, but you're right that this is a bit too long for these tests, so I've shortened it.

**Review by fenryka — APPROVED** [review]
LGTM

**Pull request merged**